// rocketmake-logger-teamcity/index.js
var rm_core = require('rocketmake');

module.exports = {
  build: (function(){
    var unhook;

    return {
      start: function(){
        console.log('##teamcity[compilationStarted compiler=\'build\']');

        unhook = rm_core.hook_stdout(function(string){
          return '##teamcity[message text=\'' + escapeString(string) + '\']\r\n';
        });
      },

      end: function(){
        unhook();
        console.log('##teamcity[compilationFinished compiler=\'build\']');
      }
    };
  })(),

  test: {
    start: function() {
      console.log('##teamcity[testSuiteStarted name=\'mstest\']');
    },
    each: function(test) {
      console.log('##teamcity[testStarted name=\'' + escapeString(test.name) + '\']');
      if(!test.passed) {
        console.log('##teamcity[testFailed name=\'' + escapeString(test.name) + '\' message=\'' + escapeString(test.errorMessage) + '\' details=\'' + escapeString(test.errorStackTrace) + '\']');
      }
      console.log('##teamcity[testFinished name=\'' + escapeString(test.name) + '\']');
    },
    end: function() {
      console.log('##teamcity[testSuiteFinished name=\'mstest\']');
    }
  },

  setVersion: function(version) {
    console.log('##teamcity[buildNumber \'' + version + '\']');
  }
};

function escapeString(string) {
  string = string || "";
  return string
  .split('|').join('||')
  .split('\'').join('|\'')
  .split('\n').join('|n')
  .split('\r').join('|r')
  .split('[').join('|[')
  .split(']').join('|]')
  .trim();
}
